app/services/admin/users_import_service.rb

```diff
@@ -0,0 +1,100 @@
+module Admin
+  class UsersImportService
+    include ActiveModel::Model
+    attr_accessor :tempfile
+
+    def initialize(*)
+      super
+      self.batch_size = 100
+      initialize_role_cache
+    end
+
+    def call
+      if tempfile.nil?
+        errors.add(:tempfile, 'No file given')
+        return
+      end
+      import_csv
+    end
+
+    private
+
+    attr_accessor :role_cache, :batch_size
+
+    def initialize_role_cache
+      self.role_cache = Hash.new do |hash, key|
+        Admin::Role.find_by(name: key).tap do |role|
+          hash[key] = role
+          raise "Unknown role #{key}" unless role.present?
+        end
+      end
+    end
+
+    def import_csv
+      lines = tempfile.readlines("\n").map { |l| l.gsub(/\n\z/, '').gsub(/\r\z/, '') }
+      rows = CSV.parse(lines.join("\n"), headers: true)
+      dups = duplicate_emails(rows)
+      if dups.present?
+        errors.add(:tempfile, "The file contains the following duplicate emails #{dups.join(' - ')}")
+      end
+      dups = duplicate_usernames(rows)
+      if dups.present?
+        errors.add(:tempfile, "The file contains the following duplicate usernames #{dups.join(' - ')}")
+      end
+      return if errors.present?
+      User.transaction do
+        begin
+          import_in_batches(rows)
+        end
+      end
+    end
+
+    def normalize_row(row)
+      {
+        username: row['username'],
+        name: row['name'],
+        email: row['email'],
+        password: row['password'],
+        password_confirmation: row['password'],
+        department: row['department'],
+        role_ids: [role_with_name(row['Role'].strip.titleize).id]
+      }
+    end
+
+    def import_in_batches(rows)
+      line_tracker = { number: 1 }
+      rows.each_slice(batch_size) do |rows|
+        begin
+          User.import import_rows(rows, line_tracker), recursive: true
+        rescue ActiveRecord::RecordInvalid => ex
+          errors.add(:tempfile, "Line #{line_tracker[:number]} - #{ex.message}")
+        end
+      end
+    end
+
+    def import_rows(rows, line_tracker)
+      batch = []
+      rows.each do |row|
+        begin
```

Comment:
I don't understand this extra block here.

Reply:
In line 65 should that be 0 instead of 1?